### PR TITLE
Attempt to fix #9919 - Intersecting discriminated union breaks type narrowing

### LINKED ISSUE #9919: Intersecting discriminated union breaks type narrowing
**TypeScript Version:**  nightly (2.0.0-dev.201xxxxx)

**Code**

```ts
type ToString = {
    toString(): string;
}

type BoxedValue = { kind: 'int',    num: number }
                | { kind: 'string', str: string }

type IntersectionFail = BoxedValue & ToString

type IntersectionInline = { kind: 'int',    num: number } & ToString
                        | { kind: 'string', str: string } & ToString

function getValueAsString(value: IntersectionFail): string {
    if (value.kind === 'int') {
        // Property 'num' does not exist on type '({ kind: "int"; num: number; } | { kind: "string"; str: string; }) & { toString(): string; }'.
        return '' + value.num;
    }

    // Property 'str' does not exist on type '({ kind: "int"; num: number; } | { kind: "string"; str: string; }) & { toString(): string; }'.
    return value.str;
}
```

If a discriminated union is enhanced via intersection, type narrowing based on the discrimination field starts to fail. This behaviour is not exhibited if a new discriminated union is created with the intersection being inlined into each of the respective options.

The function `getValueAsString(value: IntersectionFail): string` fails to compile, while `getValueAsString(value: IntersectionInline): string` compiles just fine. Unfortunately, it is `IntersectionFail` that is the result of natural program evolution.

### PR DESCRIPTION
**Pull request by gcnew**
Fixes #9919.

This PR adds support for correctly narrowing discriminated unions that are also intersected. As a side effect a type is now eligible for discrimination whenever it is a union or intersection, regardless whether the discrimination property is a union itself. This is to handle situations where the property type is narrowed by subsequent intersections.

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@ahejlsberg want to take a look?

**Comment by ahejlsberg** [comment]
@gcnew Thanks for your contribution!

We're about to merge #9407 which contains a number of changes in this area, including optimizations to improve performance in code bases with large union types (100+ constituent types) such as would result if we convert the compiler itself to depend on these features. Merging that will undoubtedly affect this PR. Not sure if you've looked at that already.

One concern I have with incorporating intersection types is that I will cause even more types to be generated by narrowing operations, thus negatively affecting performance. Also, if we go beyond just the top level of union types we really should handle any arbitrary number of nested unions and intersections, e.g. `((A | B) & (C | D) | (E | F) & (G | H))`. And of course we'd need to make sure that all the different kinds of narrowing are covered. It's definitely not a simple task.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -8095,7 +8106,7 @@ namespace ts {
                 }
                 const propName = propAccess.name.text;
                 const propType = getTypeOfPropertyOfType(type, propName);
-                if (!propType || !isStringLiteralUnionType(propType)) {
+                if (!propType || !(isStringLiteralUnionType(propType) || (type.flags & TypeFlags.UnionOrIntersection))) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: You don't want to narrow if the original type was a union? That doesn't sound right.

Comment (gcnew) [inline src/compiler/checker.ts]: Hm, maybe the better way to write it is:
```ts 
if (!propType || (!isStringLiteralUnionType(propType) && (type.flags & TypeFlags.UnionOrIntersection) === 0))
```
i.e. if neither the discrimination property is a string-literal-union, nor the parent type is union or intersection. I guess I made it unintelligible by applying De Morgan's law.

**Comment by gcnew** [comment]
@ahejlsberg Thank you for taking a look at this PR!

Yes, I'm following #9407 and I suspect that this code will not be compatible with what's to come.

I agree that complete support of arbitrary ADT/intersection blending will have memory and performance implications especially if done naively as I did here. On the other hand, this is what makes TypeScript more expressive than ML-like languages. If we can isolate the performance hit only for the _rare_ users/use cases that consciously take advantage of this feature will that be an acceptable compromise? After all having type casts and vanished type-safety is arguably worse.

**Comment by gcnew** [comment]
This behaviour is being implemented in https://github.com/Microsoft/TypeScript/pull/11717. Closing this one in favour of the latter.

**Pull request closed without merging**